modules/stochastic_tools/src/surrogates/PODReducedBasisSurrogate.C

```diff
@@ -0,0 +1,241 @@
+    {
+      _approx_solution[var_i] = DenseVector<Real>(_base[var_i][0].size());
+    }
+    _initialized = true;
+  }
+}
+
+void
+PODReducedBasisSurrogate::solveReducedSystem(const std::vector<Real> & params)
+{
+  // Cleaning the containers of the system matrix and right hand side.
+  _sys_mx.zero();
+  _rhs.zero();
+
+  // The assumption here is that the reduced operators in the trainer were
+  // assembled in the order of the parameters. Also, if the number of
+  // parameters is fewer than the number of operators, the operator will
+  // just be added without scaling.
+  for (unsigned int i = 0; i < _red_operators.size(); ++i)
+  {
+    unsigned int row_start = 0;
+
+    // Checking if the reduced operator corresponds to a Dirichlet BC, if
+    // yes introduce the penalty factor.
+    Real factor = 1.0;
+    auto it = std::find(_dir_tag_names.begin(), _dir_tag_names.end(), _tag_names[i]);
+    if (it != _dir_tag_names.end())
+      factor = _penalty;
+
+    // If the user decreased the rank of the reduced bases manually, some parts
+    // of the initial reduced operators have to be omited.
+    for (unsigned int var_i = 0; var_i < _var_names.size(); ++var_i)
+    {
+      for (unsigned int row_i = row_start; row_i < _comulative_ranks[var_i]; row_i++)
+      {
+        if (!_independent[i])
+        {
+          unsigned int col_start = 0;
+
+          for (unsigned int var_j = 0; var_j < _var_names.size(); ++var_j)
+          {
+            for (unsigned int col_i = col_start; col_i < _comulative_ranks[var_j]; col_i++)
+            {
+              if (i < params.size())
+                _sys_mx(row_i, col_i) += params[i] * factor * _red_operators[i](row_i, col_i);
+              else
+                _sys_mx(row_i, col_i) += factor * _red_operators[i](row_i, col_i);
+            }
+
+            col_start = _comulative_ranks[var_j];
+          }
+        }
+        else
+        {
+          if (i < params.size())
+            _rhs(row_i) -= params[i] * factor * _red_operators[i](row_i, 0);
+          else
+            _rhs(row_i) -= factor * _red_operators[i](row_i, 0);
+        }
+        row_start = _comulative_ranks[var_i];
+      }
+    }
+  }
+
+  // Solving the reduced system.
+  _sys_mx.lu_solve(_rhs, _coeffs);
+}
+
+void
+PODReducedBasisSurrogate::reconstructApproximateSolution()
+{
+  unsigned int counter = 0;
+  for (unsigned int var_i = 0; var_i < _var_names.size(); var_i++)
+  {
+    _approx_solution[var_i].zero();
+
+    // This also takes into account the potential truncation of the bases by
+    // the user.
+    for (unsigned int base_i = 0; base_i < _final_ranks[var_i]; ++base_i)
+    {
+      for (unsigned int dof_i = 0; dof_i < _base[var_i][base_i].size(); ++dof_i)
+        _approx_solution[var_i](dof_i) += _coeffs(counter) * _base[var_i][base_i](dof_i);
+      counter++;
+    }
+  }
+}
+
+Real
+PODReducedBasisSurrogate::getNodalQoI(std::string var_name, MooseEnum qoi_type) const
```

Comment:
Could this be the `evaluate` function instead? We can have a validParam that describes the requested QoI.

Reply:
I think we already talked about this last Friday, but I'll put the conclusion here so that Andrew can see it as well. In case of POD-RB, it doesn't make sense to create a separate surrogate for each QoI because reconstructing the approximate solution is a full-order operation (slow compared to the rest of the operations in the surrogate). So fetching multiple properties of the surrogate after reconstructing the solution is a faster way to do it. But I am open to counterarguments!